src/Integration/Aws/SimpleS3/src/SimpleS3Client.php

```diff
@@ -47,6 +53,117 @@ public function has(string $bucket, string $key): bool
+    {
+        $megabyte = 1024 * 1024;
+        if (!empty($options['ContentLength'])) {
+            $contentLength = (int) $options['ContentLength'];
+            unset($options['ContentLength']);
+        } else {
+            $contentLength = (int) $this->headObject(['Bucket' => $srcBucket, 'Key' => $srcKey])->getContentLength();
+        }
+
+        $concurrency = (int) ($options['Concurrency'] ?? 10);
+        unset($options['Concurrency']);
+        if ($concurrency < 1) {
+            $concurrency = 1;
+        }
+
+        /*
+         * The maximum number of parts is 10.000. The partSize must be a power of 2.
+         * We default this to 64MB per part. That means that we only support to copy
+         * files smaller than 64 * 10 000 = 640GB. If you are coping larger files,
+         * please set PartSize to a higher number, like 128, 256 or 512. (Max 4096).
+         */
+        $partSize = ($options['PartSize'] ?? 64) * $megabyte;
+        unset($options['PartSize']);
+
+        // If file is less than 5GB, use normal atomic copy
+        if ($contentLength < 5120 * $megabyte) {
+            $this->copyObject(
+                CopyObjectRequest::create(
+                    array_merge($options, ['Bucket' => $destBucket, 'Key' => $destKey, 'CopySource' => "{$srcBucket}/{$srcKey}"])
+                )
+            );
+
+            return;
+        }
+
+        $uploadId = $this->createMultipartUpload(
+            CreateMultipartUploadRequest::create(
+                array_merge($options, ['Bucket' => $destBucket, 'Key' => $destKey])
+            )
+        )->getUploadId();
+
+        $partNumber = 1;
+        $startByte = 0;
+        $parts = [];
+        while ($startByte < $contentLength) {
+            $parallelChunks = $concurrency;
+            $responses = [];
+            while ($startByte < $contentLength && $parallelChunks > 0) {
+                $endByte = min($startByte + $partSize, $contentLength) - 1;
+                $responses[$partNumber] = $this->uploadPartCopy(
+                    UploadPartCopyRequest::create([
+                        'Bucket' => $destBucket,
+                        'Key' => $destKey,
+                        'UploadId' => $uploadId,
+                        'CopySource' => "{$srcBucket}/{$srcKey}",
+                        'CopySourceRange' => "bytes={$startByte}-{$endByte}",
+                        'PartNumber' => $partNumber,
+                    ])
+                );
+
+                $startByte += $partSize;
+                ++$partNumber;
+                --$parallelChunks;
+            }
+            $error = null;
+            foreach ($responses as $idx => $response) {
+                try {
+                    $copyPartResult = $response->getCopyPartResult();
+                    $parts[] = new CompletedPart(['ETag' => $copyPartResult->getEtag(), 'PartNumber' => $idx]);
```

Comment:
The weird thing about the PossiblyNullReference error reported by Psalm here is that https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPartCopy.html#API_UploadPartCopy_ResponseElements documents CopyPartResult as required in the output. So it is weird that the operation was generated with a nullable type. Maybe the AWS SDK has metadata that does not match the documentation there.

Reply:
It seems there are no "required" fields for outputs

Reply:
I suggest reporting this to AWS as it looks like a mismatch between the human-readable documentation and the machine-readable data used to generate the SDKs.

Reply:
https://repost.aws/questions/QU6D3R02ZGSgG5C2-yPYw-qw/aws-php-sdk-api-2-json-is-not-synchronised-with-docs-aws-amazon-com

Reply:
what should I do?
Add additional checking? Or will wait for AWS fix?
We waiting for this functionality

Reply:
@stof any updates on this?